### PR TITLE
Threads MintsHelper

### PR DESCRIPTION
**Pull request by dgasmith**
## Description
Adds threading to the MintsHelper object. Notably we no longer build SO integrals directly in the SO basis, but build AO and transform AO->MO. This appears to be about as efficient as the former when threading, we can revisit this if the potential integrals can be improved.

Benchmark is a carbon chain in the cc-pvdz basis set, tuned to compute about a million basis functions per row.

I am separating this out from #592 which will wait for further tests.

```
# Original - 1 thread
   nbf  ntrial  ao_overlap  ao_kinetic  ao_potential  so_overlap  so_kinetic  so_potential
0   28    1275    0.002474    0.000240      0.000936    0.000253    0.000311      0.001072
1   98     104    0.005268    0.002996      0.031417    0.002182    0.002873      0.031835
2  168      35    0.007805    0.007087      0.149304    0.005561    0.007733      0.150582
3  238      17    0.019714    0.014981      0.415791    0.010558    0.014752      0.418562
4  308      10    0.029242    0.023728      0.890593    0.017159    0.023903      0.938164
5  378       6    0.038581    0.034306      1.624236    0.023958    0.034886      1.637466
6  448       4    0.058684    0.050234      2.687780    0.033669    0.048659      2.867169
7  518       3    0.073995    0.067988      4.212363    0.043464    0.063137      4.170965

# New algorithm - 1 thread
   nbf  ntrial  ao_overlap  ao_kinetic  ao_potential  so_overlap  so_kinetic  so_potential
0   28    1275    0.002221    0.000152      0.000583    0.002632    0.000274      0.000731
1   98     104    0.004644    0.001872      0.016447    0.004653    0.002240      0.016920
2  168      35    0.005474    0.004001      0.076230    0.006461    0.004780      0.078292
3  238      17    0.007382    0.006899      0.210317    0.010013    0.008758      0.214130
4  308      10    0.011478    0.011354      0.449083    0.015937    0.015050      0.452936
5  378       6    0.017539    0.016836      0.817148    0.025508    0.022129      0.825095
6  448       4    0.022457    0.023191      1.350382    0.028970    0.031135      1.363716
7  518       3    0.029087    0.030661      2.079129    0.040195    0.042085      2.086033

# New algorithm -  6 threads
   nbf  ntrial  ao_overlap  ao_kinetic  ao_potential  so_overlap  so_kinetic  so_potential
0   28    1275    0.001724    0.000046      0.001282    0.001815    0.000171      0.001468
1   98     104    0.001922    0.000262      0.004997    0.002548    0.000852      0.005569
2  168      35    0.002558    0.000697      0.019184    0.003850    0.001897      0.020388
3  238      17    0.003629    0.001342      0.049260    0.005645    0.003470      0.051254
4  308      10    0.004355    0.003332      0.120245    0.009402    0.006256      0.109358
5  378       6    0.008223    0.004361      0.190977    0.014398    0.009077      0.192586
6  448       4    0.011751    0.006664      0.311692    0.017004    0.012781      0.312895
7  518       3    0.017167    0.007853      0.475830    0.023663    0.018094      0.481414
```

As you can see efficiency is not great. We are spending about 99% of the time inside `compute_shell` however. So, if we want better efficiency were going to have to run this through `vtune` and find out whats happening in `OneBodyAOInt`. 

## Todos
Notable points that this PR has either accomplished or will accomplish.
 - [x] Moves Wavefunction classes out of `export_mints.cc`.
 - [x] Threads several common Matrix functions.
 - [x] Adds add/subtract/axpy/rms/sum_of_squares to the Vector object.
 - [x] Kills off `molden.writeNO` as this is very close to `molden.write`
 - [x] All threads now use internal P::e instead of omp_max_threads.
 - [x] NumPy now correctly compares to a `10^-precision` and not `precision`.
* **Developer Interest**
  - [x] MintsHelper is now threaded for one-body integrals.

## Questions
- [x]  It looks like `Vector::norm` actually returns sum of squares. See [here](https://github.com/psi4/psi4/compare/master...dgasmith:thread_mints?expand=1#diff-24bb65651a07c6b40a3c50b28f42449bL310). @jturney @andysim 

## Status
- [x]  Ready to go

### CODE REVIEW

File: psi4/src/psi4/libmints/mintshelper.cc [hunk 1]
```diff
@@ -161,6 +165,12 @@ void MintsHelper::common_init()
     if (print_)
         basisset_->print_detail();
 
+    // How many threads?
+    nthread_ = 1;
+    #ifdef _OPENMP
+        nthread_ = omp_get_max_threads();
```
Comment (dgasmith) [inline psi4/src/psi4/libmints/mintshelper.cc]: Need to switch this over to P::e.

**Comment by andysim** [comment]
Yeah, the norm thing is bad.  According to my QtCreator, it's not used anywhere so should be a trivial fix.

File: psi4/src/psi4/libmints/mintshelper.cc (hunk 1 above)
Comment (andysim) [inline psi4/src/psi4/libmints/mintshelper.cc]: Agreed, but make sure it still obeys the value from set_num_threads() in this class.

File: psi4/src/psi4/libmints/mintshelper.cc [hunk 2]
```diff
@@ -418,12 +428,83 @@ void MintsHelper::integral_hessians()
...
+
+    // Loop it
+    #pragma omp parallel for schedule(guided) num_threads(nthread)
+    for (size_t MU = 0; MU < bs1->nshell(); ++MU) {
+        const size_t num_mu = bs1->shell(MU).nfunction();
+        const size_t index_mu = bs1->shell(MU).function_index();
+
+        size_t rank = 0;
+#ifdef _OPENMP
+        rank = omp_get_thread_num();
```
Comment (dgasmith) [inline psi4/src/psi4/libmints/mintshelper.cc]: Thats just asking which thread it is. The one to worry about is `omp_get_max_threads`.

File: psi4/src/psi4/ccdensity/ccdensity.cc [hunk 3]
```diff
@@ -426,12 +426,14 @@ PsiReturnType ccdensity(std::shared_ptr<Wavefunction> ref_wfn, Options& options)
       if(params.write_nos){
         MoldenWriter nowriter(ref_wfn);
         std::string mol_name = ref_wfn->molecule()->name();
-        nowriter.writeNO(mol_name+"NO.molden",NOa_pair.first,NOb_pair.first,
-            NOa_pair.second,NOb_pair.second);
+        SharedMatrix NO_Ca = Matrix::doublet(Ca, NOa_pair.first, false, false);
```
Comment (dgasmith) [inline psi4/src/psi4/ccdensity/ccdensity.cc]: @amjames Look good to you?

File: psi4/src/read_options.cc [hunk 4]
```diff
@@ -179,6 +179,8 @@ int read_options(const std::string &name, Options & options, bool suppress_print
   /*- Algorithm to use for CI computation (e.g., CID or CISD).
   See :ref:`Cross-module Redundancies <table:managedmethods>` for details. -*/
   options.add_str("CI_TYPE", "CONV", "CONV");
+  /*- Write all the MOs to the MOLDEN file (true) or discard the unoccupied MOs (false). -*/
+  options.add_bool("MOLDEN_WITH_VIRTUAL", true);
```
Comment (dgasmith) [inline psi4/src/read_options.cc]: @loriab Any issue with moving this to global vars?

Comment (loriab) [inline psi4/src/read_options.cc]: No objection. And "yes, of course" on the `omp_get_thread_num()` above.

**Pull request merged**